### DISCUSSION TITLE
[FEATURE REQUEST] - Switch the default sampler to k_euler

### DISCUSSION THREAD

**Discussion by altryne** [hlky/stable-diffusion-webui#277] [Feature suggestions]
**Is your feature request related to a problem? Please describe.**
Suggestion to switch the default sampler to k_euler and the default number of steps to 15-30 

For most folks this would result in faster but still coherent results

**Comment by oobabooga** [comment]
I would suggest `k_euler_a` with 12 steps, which works very nicely for me. 12-18 steps seems to be a sweet spot.

**Reply by Kallamamran** [reply]
Never tried that one, but beg to agree. That was a very good diffusion model

**Comment by hlky** [comment]
@altryne @oobabooga can you post some examples? we should consider more options as well

**Comment by hlky** [comment]
easy to change in [the UI defaults](https://github.com/hlky/stable-diffusion-webui/blob/master/webui.yaml) though

**Comment by ZeroCool940711** [comment]
@hlky check this discussion #178.

**Comment by hlky** [comment]
@ZeroCool940711 I saw those, @JoshuaKimsey posted it after my wiki page for [benchmarking](https://github.com/hlky/stable-diffusion/wiki/Benchmarking) which functions as a comparison too.

Clearly what I meant was examples of the specific settings that @altryne @oobabooga think the default UI settings should be changed to.

**Comment by ZeroCool940711** [comment]
Oh, sorry I misunderstood what you were asking for, my bad

**Comment by oobabooga** [comment]
I don't have examples that objectively prove a point, but what I can tell is that `k_euler_a` gives satisfactory results with 12 steps, while taking me 30-40 seconds per image vs 2 minutes for the current defaults. 

Before that, I was using DDIM with 50 steps and I did not perceive any change in quality.

This is really anecdotal, but it helped me save some time on my humble 4GB GPU.

**Comment by hlky** [comment]
Maybe a solution would be to provide a dropdown of set configurations of sampler and steps, like 'very fast', 'fast', 'slow', 'very slow'

**Comment by ZeroCool940711** [comment]
I think a better solution would be to make it so we can save our own default values, like a button to save the defaults to a file, that way if we think some settings work better for us or there are some settings we use a lot we can save it and that will be the default for us, everyone might have different opinion on what is best or things might work different on their computer because of their hardware, been able to save our own defaults would solve this problem in a better way than changing it for everyone every now and then.

**Comment by hlky** [comment]
For now, just edit [ui defaults](https://github.com/hlky/stable-diffusion-webui/blob/master/webui.yaml)
I agree saving the defaults would be a good feature though

**Reply by altryne** [reply]
Agree, I'be been missing this feature myself!

**Comment by oobabooga** [comment]
An additional button is unnecessary and would be bloat imo. The ui defaults work fine for setting your personal preferences. 

The only benefit in changing the defaults would be to help users who do not know what they are doing.

**Comment by JoshuaKimsey** [comment]
I would personally recommend against using the `_a` variants as a default, simply because the results they produce will be wildly different from what most prompt crafters and references will show for a given seed and setup. That said, I would be ok with using the `k_euler` as a default, given its results are fairly similar to the current community default of `k_lms`. `k_lms` definitely is considered by many to be the best default for Stable Diffusion, but we are only a week into this, so things may change towards viewing `k_euler` as the best default instead.

**Reply by darkstorm2150** [reply]
k_euler is fast at 25 steps and 4 cfg, its coherent with k_lms results.

Tested these results on it


![index](https://user-images.githubusercontent.com/3039752/187538312-6b427c0e-28bd-4d42-bcef-ecae5d845392.png)


PROMPT: Kenshi game, robotic ultra thin body, robotic machine, mechanical face, lens eyes, intense fighting with epic swords against rebel farmers, limbs flying, blood, insanely detailed by Greg Rutkowski and Claude Monet, oil on canvas
Steps: 18, Sampler: k_euler_a, CFG scale: 7.5, Seed: 700320420

18 steps works!

**Reply by darkstorm2150** [reply]
This is with k_euler

![index2](https://user-images.githubusercontent.com/3039752/187538764-59000445-f08d-478e-9e24-992f06efd918.png)

PROMPT: Kenshi game, robotic ultra thin body, robotic machine, mechanical face, lens eyes, intense fighting with epic swords against rebel farmers, limbs flying, blood, insanely detailed by Greg Rutkowski and Claude Monet, oil on canvas
Steps: 18, Sampler: k_euler, CFG scale: 7.5, Seed: 2695351964

**Reply by hlky** [reply]
I think defaults of 20-30 steps k_euler cfg scale 7.5 would work and produce good results faster than the current defaults. k_lms did use slightly less vram for me though and right now trying to support as many users as possible as easily as possible with the limitations of common hardware is just as important as speed. I wouldn't want someone getting discouraged from creating if the defaults were changed then the extra vram affected some users, and changing the sampler becomes another thing to the list of troubleshooting steps for low end hardware users.

**Reply by JoshuaKimsey** [reply]
Hmm, interesting. I just looked back at my Benchmarks and k_euler actually used less VRAM for me? Not by a huge margin, but about a percentage point. Could just be random noise though. Would probably need to do 10 renders on both and then average it out to know for sure.

**Comment by darkstorm2150** [comment]
Also this was added on discord for Stable Diffusion

********************************
Technical details regarding Stable Diffusion samplers, confirmed by Katherine:
    - DDIM and PLMS are originally the Latent Diffusion repo
        ○ DDIM was implemented by CompVis group and was default (slightly different update rule than the samplers below, eqn 15 in DDIM paper is the update rule vs solving eqn 14's ODE directly)
        ○ PLMS was implemented by Katherine a while back and added, sped up convergence from like 250 steps to 50 steps
    - All k_ samplers are taken from Katherine's k-diffusion repository
    - k_euler_ancestral is ancestral sampling with Euler's (or technically Euler-Maruyama) method from the variance-exploding SDE for a DDPM
    - k_euler is sampling with Euler's method from the DDIM probability flow ODE
    - k_heun is sampling with Heun's method (2nd order method, recommended by Karras et al.) from the DDIM probability flow ODE
    - k_lms is sampling with linear multi-step method (4th-order Adams-Bashforth, first step 1st order Euler, second step 2nd order Heun, etc. till 4th step, then subsequently depending on the past 4 steps) of the DDIM probability flow ODE
    - k_lms_ancestral - DDPM as an ODE with k_lms then add noise
    - k_dpm_2 - implementation of DPM solver paper with the DDIM probability flow ODE with some additional improvements by Katherine
    - k_dpm_2_ancestral - k_dpm_2 with DDPM as an ODE then add noise

k_dpm_2_ancestral is Katherine's favorite.

Katherine is also puzzled why k_euler_ancestral is doing so well as that's the least efficient sampler.

She mentioned that for diffusion directly on the pixels would give very blurry results for k_euler_ancestral, which she's seen for many generations.

Her hypothesis is that the autoencoder in latent diffusion is leading to completely different dynamics where even if the samples aren't best in the latent space, the autoencoder is able to map it to a good image. (edited)

********************************

**Reply by hlky** [reply]
In my [Benchmarking](https://github.com/hlky/stable-diffusion/wiki/Benchmarking) I tested the same seed and prompt with all the samplers, the results with k_euler_ancestral and k_dpm_2_ancestral were interesting, different than the others but similar to each other.

**Reply by JoshuaKimsey** [reply]
Yeah, the `_a` results tend to be really good at low sample rates, but the issue is that they are so very different from the other samplers. That said, I do support k_euler being made the default sampler, as it does seem to produce better results at lower steps, at least compared to k_lms.

**Reply by andrewginns** [reply]
For reference here's a comparison.

https://i.ibb.co/vm4fm7L/1661440027115223.jpg
